File: categoryTheory/Chapter2/memoize.gpt.cpp
#include <functional>
#include <unordered_map>

template <typename T, typename R>
std::function<R(T)> memoize(std::function<R(T)> fn) {
    std::unordered_map<T, R> cache;
    return [fn, cache](T t) mutable -> R {
        auto it = cache.find(t);
        if (it != cache.end()) {
            return it->second;
        }
        R result = fn(t);
        cache[t] = result;
        return result;
    };
}

File: categoryTheory/Chapter2/memoize.gpt.rs
use std::collections::HashMap;
use std::hash::Hash;

fn memoize<T: Eq + Hash, R, F: Fn(T) -> R>(fn: F) -> impl Fn(T) -> R {
    let mut cache = HashMap::new();

    move |t: T| -> R {
        if let Some(result) = cache.get(&t) {
            *result
        } else {
            let result = fn(t);
            cache.insert(t, result);
            result
        }
    }
}
